engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/TypeOfNode.java

```diff
@@ -193,4 +103,169 @@ Object doAny(Object value) {
+    }
+
+    @Specialization(guards = {"findInterop(value, interop).isString()"})
+    Type doPolyglotString(Object value, @CachedLibrary(limit = "3") InteropLibrary interop) {
+      return EnsoContext.get(this).getBuiltins().text();
+    }
+
+    @Specialization(guards = {"findInterop(value, interop).isNumber()"})
+    Type doPolyglotNumber(Object value, @CachedLibrary(limit = "3") InteropLibrary interop) {
+      Builtins builtins = EnsoContext.get(this).getBuiltins();
+      if (interop.fitsInInt(value)) {
+        return builtins.number().getInteger();
+      } else if (interop.fitsInDouble(value)) {
+        return builtins.number().getDecimal();
+      } else {
+        return EnsoContext.get(this).getBuiltins().number().getNumber();
+      }
+    }
+
+    @Specialization(guards = {"findInterop(value, interop).isDateTime()"})
+    Type doDateTime(Object value, @CachedLibrary(limit = "3") InteropLibrary interop) {
+      return EnsoContext.get(this).getBuiltins().dateTime();
+    }
+
+    @Specialization(guards = {"findInterop(value, interop).isTimeZone()"})
+    Type doTimeZone(Object value, @CachedLibrary(limit = "3") InteropLibrary interop) {
+      EnsoContext ctx = EnsoContext.get(this);
+      return ctx.getBuiltins().timeZone();
+    }
+
+    @Specialization(guards = {"findInterop(value, interop).isDate()"})
+    Type doDate(Object value, @CachedLibrary(limit = "3") InteropLibrary interop) {
+      EnsoContext ctx = EnsoContext.get(this);
+      return ctx.getBuiltins().date();
+    }
+
+    @Specialization(guards = {"findInterop(value, interop).isTime()"})
+    Type doTime(Object value, @CachedLibrary(limit = "3") InteropLibrary interop) {
+      EnsoContext ctx = EnsoContext.get(this);
+      return ctx.getBuiltins().timeOfDay();
+    }
+
+    @Specialization(guards = "findInterop(value, interop).isDuration()")
+    Type doDuration(Object value, @CachedLibrary(limit = "3") InteropLibrary interop) {
+      EnsoContext ctx = EnsoContext.get(this);
+      return ctx.getBuiltins().duration();
+    }
+
+    @Specialization(guards = {"findInterop(value, interop).isMetaObject()"})
+    Object doMetaObject(Object value, @CachedLibrary(limit = "3") InteropLibrary interop) {
+      try {
+        return interop.getMetaObject(value);
+      } catch (UnsupportedMessageException e) {
+        CompilerDirectives.transferToInterpreter();
+        Builtins builtins = EnsoContext.get(this).getBuiltins();
+        throw new PanicException(builtins.error().makeCompileError("invalid meta object"), this);
+      }
+    }
+
+    @Fallback
+    @CompilerDirectives.TruffleBoundary
+    Object doAny(Object value) {
+      return DataflowError.withoutTrace(
+          EnsoContext.get(this)
+              .getBuiltins()
+              .error()
+              .makeCompileError("unknown type_of for " + value),
+          this);
+    }
+
+    static InteropType findInterop(Object value, InteropLibrary interop) {
```

Comment:
```suggestion
    static InteropType findInteropType(Object value, InteropLibrary interop) {
```

Wouldn't that name be cleaner?

What it does is finds an `InteropType`, `findInterop` suggested it finds some `Interop`.

Reply:
Shortened and renamed in f32e2ba